### ISSUE TITLE
Storage

### ISSUE THREAD

**Issue by perguth** [peermusic/research#1]
How to store large amounts of data in the browser?

Related user stories:
- [[6] Als Entwickler moechte ich ueber client-seitige Datenbanken (Android) bescheid wissen.](https://github.com/peermusic/user-stories/issues/3)
- [[4] Als Entwickler moechte ich ueber client-seitige Datenbanken (Desktop) bescheid wissen.](https://github.com/peermusic/user-stories/issues/2)

---

Consolidate knowledge here: [storage.md](https://github.com/peermusic/research/blob/master/storage.md)

**Comment by perguth** [comment]
There is a good overview on [HTML5 Rocks](http://www.html5rocks.com/en/features/storage).

**Comment by queicherius** [comment]
There is a some research on quotas here: http://www.html5rocks.com/en/tutorials/offline/quota-research/

Also Chrome official documentation here https://developer.chrome.com/apps/offline_storage#persistent

**Comment by queicherius** [comment]
After playing around a bit with http://demo.agektmr.com/storage/ here are the options to save > 1GB of data permanently, with quota permissions:
- localStorage: Nope
- SessionStorage: Nope
- WebSQL: Was making me lag really badly and crashed my browser multiple times.
- IndexedDB: Was making me lag, really weird behaviour with permanent/temp storage. Really slow.
- **FileSystem: Worked without complaints. Way faster than the other options. I'd suggest we use this (already suggested by @ryomiyaichi as well).**

The drawbacks to FileSystem are, that it only works on Chrome, but after reading up a bit on it, I don't think we have a good shot with other browsers anyway...

**Comment by ryomiyaichi** [comment]
Overview of technics that **@queicherius** mentioned: http://www.html5rocks.com/en/tutorials/offline/storage/

**Comment by perguth** [comment]
Results here: https://github.com/peermusic/research/blob/master/music-recognition.md

**Comment by perguth** [comment]
Regarding FileSystem API [MDN](https://developer.mozilla.org/en-US/docs/Web/API/FileSystem) writes:

> "This feature is non-standard and is **not on a standards track**. [...] This document was last updated on March 2, 2012 and follows the W3C Specifications (Working Draft) drafted on April 19, 2011.
> The draft is no more actively maintained, the **specification is abandonned** for the moment as it didn't get any significant traction."

Because of this I did a small test using IndexedDB wrapped by [LevelUP](https://github.com/Level/levelup). LevelUP can utilize [Level-js](https://github.com/maxogden/level.js) which states as one of their goals: "Store large amounts of ascii (strings, JSON) and binary (ArrayBuffers, Typed Arrays) data in modern browsers" 

Code is here: https://github.com/peermusic/file-system/tree/levelup

To run:

``` sh
npm install -g wzrd
npm install level-browserify
wzrd indexeddb.js
```

I ~~created/stored 50x 50MB blobs~~ which persisted over restarts of the browser. I currently only store strings. Will fix that in a bit.

**Comment by queicherius** [comment]
Is this closed with f94555fe8bc48c27ca717ce40b4c8731119a586a?

**Comment by perguth** [comment]
Yeah, thanks!
